### ISSUE TITLE
底层卷积函数的疑惑

### ISSUE THREAD

**Issue by YuBicheng** [YunYang1994/tensorflow-yolov3#244]
在阅读您源码的过程中发现了一疑惑。
在core.common 中convolutional函数有一个downsample参数，在阅读源码后发现，这个参数的输入为真时，虽然最后使用tensorflow的卷积函数时会将padding参数设置为'VALID'，也就是不进行填充，但是在这之前会手动计算需要填充的边框大小并且使用tf.pad()函数填充边框。是不是我的代码读错了？
以下是您的源码：

`def convolutional(input_data, filters_shape, trainable, name, downsample=False, activate=True, bn=True):

    with tf.variable_scope(name):
        # TODO(问题待解决) 下采样为真时会使用人工填充的方式补充padding,在卷积时不使用函数自带的padding
        if downsample:
            # 计算填充的行数和列数
            pad_h, pad_w = (filters_shape[0] - 2) // 2 + 1, (filters_shape[1] - 2) // 2 + 1
            # 这两行是给输入数据外填充的，输入数据的维度为[batchsize, 特征图宽度, 特征图高度, 特征图深度]
            # 当为第一层卷积时特征图就是原图片
            # tf.pad行数会将对应维度添加padding
            # 如下 是对batch维度不添加padding 宽高维度添加对应padding 图片深度不添加padding
            paddings = tf.constant([[0, 0], [pad_h, pad_h], [pad_w, pad_w], [0, 0]])
            input_data = tf.pad(input_data, paddings, 'CONSTANT')
            # 步长
            strides = (1, 2, 2, 1)
            padding = 'VALID'
        else:
            strides = (1, 1, 1, 1)
            padding = "SAME"

        weight = tf.get_variable(name='weight', dtype=tf.float32, trainable=True,
                                 shape=filters_shape, initializer=tf.random_normal_initializer(stddev=0.01))
        conv = tf.nn.conv2d(input=input_data, filter=weight, strides=strides, padding=padding)

        if bn:
            conv = tf.layers.batch_normalization(conv, beta_initializer=tf.zeros_initializer(),
                                                 gamma_initializer=tf.ones_initializer(),
                                                 moving_mean_initializer=tf.zeros_initializer(),
                                                 moving_variance_initializer=tf.ones_initializer(), training=trainable)
        else:
            bias = tf.get_variable(name='bias', shape=filters_shape[-1], trainable=True,
                                   dtype=tf.float32, initializer=tf.constant_initializer(0.0))
            conv = tf.nn.bias_add(conv, bias)

        if activate == True: conv = tf.nn.leaky_relu(conv, alpha=0.1)

    return conv`

**Comment by thunderVolibear** [comment]
我也困惑这个问题。tf.nn.conv2d函数自带有padding = 'VALID'这个设置选项，为什么前面还要用tf.pad()手动填充一次。这样不是填充了两次吗？

**Comment by YuBicheng** [comment]
这个问题已经解决了！！！我眼瞎看错了，傻瓜有我一个就行了，仔细看代码就明白咋回事了，逻辑 代码都没有任何问题。
